src/net/rpc.rs

```diff
@@ -332,14 +351,147 @@ impl ConversationHttp {
+                                      contract_addr: &StacksAddress, contract_name: &ContractName, map_name: &ClarityName, key: &Value) -> Result<(), net_error> {
+        let response_metadata = HttpResponseMetadata::from(req);
+        let contract_identifier = QualifiedContractIdentifier::new(contract_addr.clone().into(), contract_name.clone());
+
+        let data = chainstate.with_read_only_clarity_tx(cur_burn, cur_block, |clarity_tx| {
+            clarity_tx.with_clarity_db_readonly(|clarity_db| {
+                let key = ClarityDatabase::make_key_for_data_map_entry(&contract_identifier, map_name, key);
+                let (value, marf_proof) = clarity_db.get_with_proof::<Value>(&key)
+                    .map(|(a, b)| (a, b.to_hex()))
+                    .unwrap_or_else(|| (Value::none(), "".into()));
+                let data = value.serialize();
+                MapEntryResponse { data, marf_proof }
+            })
+        });
+
+        let response = HttpResponseType::GetMapEntry(
+            response_metadata, data);
+
+        response.send(http, fd).map(|_| ())
+    }
+
+    fn handle_readonly_function_call<W: Write>(http: &mut StacksHttp, fd: &mut W, req: &HttpRequestType,
+                                               chainstate: &mut StacksChainState, cur_burn: &BurnchainHeaderHash,
+                                               cur_block: &BlockHeaderHash, contract_addr: &StacksAddress, contract_name: &ContractName,
+                                               function: &ClarityName, sender: &PrincipalData, args: &[Value] ) -> Result<(), net_error> {
+        let response_metadata = HttpResponseMetadata::from(req);
+        let contract_identifier = QualifiedContractIdentifier::new(contract_addr.clone().into(), contract_name.clone());
+
+        let cost_track = LimitedCostTracker::new(ExecutionCost { write_length: 0, write_count: 0,
+                                                                 read_length: 100000, read_count: 10,
+                                                                 runtime: 10000000 });
+
+        let args: Vec<_> = args.iter().map(|x| SymbolicExpression::atom_value(x.clone())).collect();
+
+        let data = chainstate.with_read_only_clarity_tx(cur_burn, cur_block, |clarity_tx| {
+            clarity_tx.with_readonly_clarity_env(sender.clone(), cost_track, |env| {
+                env.execute_contract(&contract_identifier, function.as_str(), &args, true)
+            })
+        });
+
+
+        let response = match data {
+            Ok(data) => HttpResponseType::CallReadOnlyFunction(response_metadata, data.serialize()),
+            Err(e) => {
+                let mut err_msg = HashMap::new();
+                err_msg.insert("message".into(), "Error executing read only function".into());
+                err_msg.insert("cause".into(), e.to_string());
+                HttpResponseType::BadRequestJSON(response_metadata, err_msg)
```

Comment:
Why is this an error?  The VM finished the execution without issue.  Shouldn't this be a 200, and shouldn't the success/failure (more generally: the return value) be part of the returned HTTP body?

Reply:
Generally speaking, I'm inclined to only use HTTP error codes to indicate problems with the request itself or unrecoverable errors in the server.  If the application behind the HTTP server does the right thing (even if, from the application's perspective, the result is interpreted as an error), then the HTTP server should return 200.

Reply:
This kind of error will necessarily lead to a different JSON object than an actual result (the error isn't a Clarity value at all, rather it's some kind of Rust error object). I think it simplifies client error handling if they can use HTTP error codes to trigger the error handling, but ultimately it doesn't really matter beyond creating a more complex JSON schema for 200 responses.

Reply:
I'm thinking longer term.  CDNs and middleboxes should cache error-responses from Clarity, since an error-response from a smart contract is still a valid response.  If you're worried about client complexity, then maybe the schema here should something like:

```
{
   "success": "boolean",
   "value": { /* clarity value schema */ }
}
```

Reply:
CDNs are perfectly happy to cache 4xx responses. It's 5xx responses that they'll mostly refuse to cache. That being said, I don't think this is a particularly important detail, so I'm happy to change these to 200 errors.

Reply:
Updated in 748a98b40ee2c0bd53272e5e468bf180074573b8